### PR TITLE
Exclude successful response content from the message of RequestFailedException

### PR DESCRIPTION
**Pull request by christothes**
Do not add response content to the exception message if the response was successful.

### CODE REVIEW

**Review by JoshLove-msft — APPROVED** [review]

File: sdk/core/Azure.Core/src/RequestFailedException.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/RequestFailedException.cs:24]: nit: 

```suggestion
        internal const string NoContentOnSuccessMessage = "Service request succeeded. Response content is not included to avoid logging sensitive data.";
```

File: sdk/core/Azure.Core/src/RequestFailedException.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/RequestFailedException.cs:178]: Why not include ReasonPhrase with a success result?

File: sdk/core/Azure.Core/src/RequestFailedException.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/RequestFailedException.cs:220]: Curious: is it a security risk to include sanitized headers in the exception message?  I don't feel strongly one way or another, but my initial understanding of the proposal was that only the response body would be omitted from the exception message, so just looking to get my head around the delta from that mental model.

File: sdk/core/Azure.Core/src/RequestFailedException.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/RequestFailedException.cs:176]: nit: consider refactoring to call helper routines in favor of a big if/else block?

**Review by annelo-msft — COMMENTED** [review]
One non-blocking philosophical question that you are free to ignore -- if the request didn't fail, should we throw a different exception than RequestFailedException?

File: sdk/core/Azure.Core/src/RequestFailedException.cs
Comment (annelo-msft) [reply sdk/core/Azure.Core/src/RequestFailedException.cs:24]: I believe response content usually means the HTTP body - if we remove more than just the content from the exception message, we might want to change this to reflect that.

File: sdk/core/Azure.Core/src/RequestFailedException.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/RequestFailedException.cs:178]: I debated whether to include it - what is a typical value for success?

Comment (annelo-msft) [reply sdk/core/Azure.Core/src/RequestFailedException.cs:178]: I think it's part of the HTTP spec, so it seems strange to remove.  It's like 200: "OK", 404: "NotFound", stuff like that.

File: sdk/core/Azure.Core/CHANGELOG.md
Comment (christothes) [reply sdk/core/Azure.Core/CHANGELOG.md:12]: Yes, I think we should add it there also.

File: sdk/core/Azure.Core/src/RequestFailedException.cs
Comment (annelo-msft) [reply sdk/core/Azure.Core/src/RequestFailedException.cs:178]: But if you know of a case where services rewrite it with sensitive content, that would be a reason to remove it.  I just don't know of one.

File: sdk/core/Azure.Core/src/RequestFailedException.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/RequestFailedException.cs:176]: Wouldn't we just push the big if/else to the helper routines? Or did you have something specific in mind to refactor?

File: sdk/core/Azure.Core/src/RequestFailedException.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/RequestFailedException.cs:220]: Unless you feel strongly, I prefer to redact everything but the status. The raw response is always available to inspect if needed.

File: sdk/core/Azure.Core/src/RequestFailedException.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/RequestFailedException.cs:178]: makes sense, I'll add it back.

File: sdk/core/Azure.Core/CHANGELOG.md
Comment (christothes) [reply sdk/core/Azure.Core/CHANGELOG.md:12]: Looking into this, we already don't log the content in `ClientResultException`

File: sdk/core/Azure.Core/src/RequestFailedException.cs
Comment (annelo-msft) [reply sdk/core/Azure.Core/src/RequestFailedException.cs:220]: I don't feel strongly, and happy to defer to you -- and also more conservative/safer approaches in general -- on security topics :)

File: sdk/core/Azure.Core/CHANGELOG.md
Comment (annelo-msft) [reply sdk/core/Azure.Core/CHANGELOG.md:12]: Ah, fair.  I had thought we were logging the headers, but I just looked and we are not currently.

As a note we do have this outstanding request to add content logging: https://github.com/Azure/azure-sdk-for-net/issues/42823

It would be interesting to think about what we do there in light of this change, and if we were to implement inheritance of `RequestFailedException` from `ClientResultException`, could we easily implement the more secure logic in one place so we didn't get accidental drift there somehow.  Dunno how much of that you're interested in proactively solving as part of this change or leaving to future work -- we certainly don't need to block this PR on solutions to that problem, but if you spot good solutions easily, it might be nice to document them on that issue or elsewhere for whoever takes that item in later work.

File: sdk/core/Azure.Core/src/RequestFailedException.cs
Comment (annelo-msft) [reply sdk/core/Azure.Core/src/RequestFailedException.cs:176]: In the push for code clarity in Azure.Core 2.0 we worked to refactor big if/else statements to something like 

```csharp
if (condition)
{
  A();
}
else
{
  B();
}
```

or similar, so it's easy to conceptualize the high-level algorithm at a glance and then drill into further details in the leaf-nodes of the function call tree.  "Programming by intention" I think it is sometimes called?

In this case, it might be a little more clear to do something like:

```csharp
messageBuilder.Append( response.IsError ? GetErrorResponseMessage() : GetSuccessResponseMessage() );
```

or 

```csharp
if (response.IsError)
{
  AppendErrorResponseMessage(messageBuilder);
}
else
{
  AppendSanitizedSuccessResponseMessage(messageBuilder);
}
```

But it's a fine point, so your call in the end.

File: sdk/storage/Azure.Storage.DataMovement/src/JobPartInternal.cs
Comment (christothes) [inline sdk/storage/Azure.Storage.DataMovement/src/JobPartInternal.cs:389]: @amnguye In this case we were actually looking for the ErrorCode in the case of `BlobAlreadyExists`, and the contents of the full response content wasn't actually required. Would this cause any other problems for other values of [DataMovementConstants.ErrorCode.CannotOverwrite](https://github.com/Azure/azure-sdk-for-net/blob/ef9e6bb7d125031fbb1a7ed25521eaa4143b9f48/sdk/storage/Azure.Storage.DataMovement/src/Shared/DataMovementConstants.cs#L142C68-L142C85)?

Comment (jsquire) [reply sdk/storage/Azure.Storage.DataMovement/src/JobPartInternal.cs:389]: //cc: @schaabs

**Review by jsquire — APPROVED** [review]

File: sdk/storage/Azure.Storage.DataMovement/tests/CleanUpTransferTests.cs
Comment (christothes) [inline sdk/storage/Azure.Storage.DataMovement/tests/CleanUpTransferTests.cs:127]: This test and the one below seemed to be asserting incorrectly for calling `DeleteIfExistsAsync` based on the comment [here](https://github.com/Azure/azure-sdk-for-net/blob/ef9e6bb7d125031fbb1a7ed25521eaa4143b9f48/sdk/storage/Azure.Storage.DataMovement/src/JobPartInternal.cs#L453-L455). Can you confirm @amnguye ?

File: sdk/core/Azure.Core/CHANGELOG.md
Comment (christothes) [reply sdk/core/Azure.Core/CHANGELOG.md:12]: I think linking to this PR is good enough reference for now. I'd be happy to pick up the other work after this.

File: sdk/storage/Azure.Storage.DataMovement/src/JobPartInternal.cs
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement/src/JobPartInternal.cs:389]: Correct we are just looking for `BlobAlreadyExists` (and possibly other error codes) in the future. So this wouldn't cause problems for the other cannot overwrite values.

File: sdk/storage/Azure.Storage.DataMovement/tests/CleanUpTransferTests.cs
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement/tests/CleanUpTransferTests.cs:127]: Let me look into this. I would have expected for this not to change based on the change you made.

Because the test doesn't have overwrite enabled, `DeleteIfExistsAsync` should be still called here since the error was `AuthenticationFailed` in this specific test.

Comment (christothes) [reply sdk/storage/Azure.Storage.DataMovement/tests/CleanUpTransferTests.cs:127]: The reason I think that it may have been wrong before is that according to [this logic](https://github.com/Azure/azure-sdk-for-net/blob/ef9e6bb7d125031fbb1a7ed25521eaa4143b9f48/sdk/storage/Azure.Storage.DataMovement/src/JobPartInternal.cs#L599), we should only call `DeleteIfExistsAsync` when the job failure type is `JobPartFailureType.Other`.

Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement/tests/CleanUpTransferTests.cs:127]: Yes you're right. We shouldn't be calling `DeleteIfExistsAsync` in this case. Especially since it's a authentication failure, calling `DeleteIfExistsAsync` would also be pointless here since we can't authenticate anyways to call delete.

Thanks for correcting this.

**Review by amnguye — APPROVED** [review]

File: sdk/core/Azure.Core/src/Pipeline/RequestFailedDetailsParser.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/Pipeline/RequestFailedDetailsParser.cs:18]: Super nit:

```suggestion
        /// In various scenarios, parsers may be called for successful responses. Implementations should not populate <paramref name="error"/> or <paramref name="data"/> with sensitive information, as these values may be logged as part of the exception.
```

But otherwise, ❤️ it! :)

File: sdk/core/Azure.Core/src/RequestFailedException.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/RequestFailedException.cs:171]: Nice rewrite, and thank you! 👏

**Review by annelo-msft — APPROVED** [review]

File: sdk/core/Azure.Core/src/Pipeline/RequestFailedDetailsParser.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/Pipeline/RequestFailedDetailsParser.cs:18]: thanks!

**Pull request merged**
